### PR TITLE
change the logic on handling configuration change

### PR DESCRIPTION
**Pull request by pan-wang**
listen on global config change notification and check whether there is any application needs to be recycled.
Recycle will call  application shutdown first instead of rely on destructor which has issue if there is long running request (i.e., reference counter will not reach to zero and process stop will not be called). Make the application shutdown on a different thread to avoid locking the application manager hash table for long time.

### CODE REVIEW

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 1]
```diff
@@ -197,72 +222,194 @@ APPLICATION_MANAGER::RecycleApplication(
...
 
-    if(dwPreviousCounter != m_pApplicationInfoHash->Count())
+    if (dwPreviousCounter != m_pApplicationInfoHash->Count())
     {
-        // Application got recycled. Log an event
-        STACK_STRU(strEventMsg, 256);
-        if (SUCCEEDED(strEventMsg.SafeSnwprintf(
-            ASPNETCORE_EVENT_RECYCLE_CONFIGURATION_MSG,
-            pszApplicationId)))
+        if (m_hostingModel == HOSTING_IN_PROCESS)
```
Comment (bangbingsyb) [inline src/AspNetCore/Src/applicationmanager.cxx]: For in-proc scenario, it looks like we don't actually need to perform the shallow copy of the hashtable? Since we don't need to modify m_pApplicationInfoHash at the end, all we need to know is whether the configure change affects one of the applications in the hash table, and then we only need to recycle w3wp upon that? Maybe we could have a small optimization here:?

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 2]
```diff
@@ -197,72 +222,194 @@ APPLICATION_MANAGER::RecycleApplication(
...
 
     if (m_pApplicationInfoHash->Count() == 0)
     {
         m_hostingModel = HOSTING_UNKNOWN;
     }
 
+    ReleaseSRWLockExclusive(&m_srwLock);
 
-    if (g_fAspnetcoreRHLoadedError)
+    if(!context.MultiSz.IsEmpty() && (m_hostingModel == HOSTING_OUT_PROCESS || m_hostingModel == HOSTING_UNKNOWN))
```
Comment (bangbingsyb) [inline src/AspNetCore/Src/applicationmanager.cxx]: For out-proc, we do update m_pApplicationInfoHash within the lock.
What would be the behavior of the requests coming to the app right after the table update?
What would be the behavior of the requests being processed in the backend app? Within shutdown limit, if their responses come, does ANCM still forward them back to clients?

File: src/AspNetCore/Src/applicationinfo.cpp [hunk 3]
```diff
@@ -132,6 +131,8 @@ APPLICATION_INFO::UpdateAppOfflineFileHandle()
             }
         }
 
+        m_fAppOfflineFound = TRUE;
```
Comment (jkotalik) [inline src/AspNetCore/Src/applicationinfo.cpp]: Why was this changed?

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 4]
```diff
@@ -189,6 +189,31 @@ APPLICATION_MANAGER::GetApplicationInfo(
     return hr;
 }
 
+BOOL
+APPLICATION_MANAGER::FindConfigChangedApplication(
```
Comment (jkotalik) [inline src/AspNetCore/Src/applicationmanager.cxx]: I'd prefer to have this as a utility function as it doesn't depend on anything else in the application manager.

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 5]
```diff
@@ -197,72 +222,194 @@ APPLICATION_MANAGER::RecycleApplication(
...
+
+    APPLICATION* pApplication = pEntry->QueryApplication();
+
+    // Reference the application first
+    pApplication->ReferenceApplication();
+
+    // Reset application pointer to NULL
+    // The destructor of ApplictionInfo will not call ShutDown again
+    pEntry->ResetApplication();
+    HANDLE hThread = CreateThread(
```
Comment (jkotalik) [inline src/AspNetCore/Src/applicationmanager.cxx]: Don't we need to wait on this thread to finish before calling close handle?

File: src/CommonLib/resources.h [hunk 6]
```diff
@@ -28,8 +28,9 @@
...
 #define ASPNETCORE_EVENT_INPROCESS_THREAD_EXIT_STDOUT_MSG    L"Application '%s' with physical root '%s' hit unexpected managed background thread exit, ErrorCode = '0x%x. Last 4KB characters of captured stdout and stderr logs:\r\n%s"
 #define ASPNETCORE_EVENT_INPROCESS_THREAD_EXIT_MSG           L"Application '%s' with physical root '%s' hit unexpected managed background thread exit, ErrorCode = '0x%x. Please check the stderr logs for more information."
-#define ASPNETCORE_EVENT_RECYCLE_APPOFFLINE_MSG              L"Application '%s' is recycled due to app_offline file was detected."
-#define ASPNETCORE_EVENT_RECYCLE_CONFIGURATION_MSG           L"Application '%s' is recycled due to configuration changed"
+#define ASPNETCORE_EVENT_RECYCLE_APPOFFLINE_MSG              L"Application '%s' is recycled due to app_offline file."
+#define ASPNETCORE_EVENT_RECYCLE_CONFIGURATION_MSG           L"Application '%s' is recycled due to configuration change"
+#define ASPNETCORE_EVENT_RECYCLE_FAILURE_CONFIGURATION_MSG   L"Failed to recyle application for configuration change and had to recycle IIS worker process, Error Code = '0x%x'."
```
Comment (jkotalik) [inline src/CommonLib/resources.h]: nit: "Failed to recycle application due to configuration change and had to recycle IIS worker process."

File: src/RequestHandler/outofprocess/outprocessapplication.cpp [hunk 7]
```diff
@@ -57,10 +57,14 @@ __override
 VOID
 OUT_OF_PROCESS_APPLICATION::ShutDown()
 {
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/RequestHandler/outofprocess/outprocessapplication.cpp]: I believe that is the case.

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 8]
```diff
@@ -197,72 +222,194 @@ APPLICATION_MANAGER::RecycleApplication(
...
+            // clean up the hash table so that the application will be informed on shutdown
+            m_pApplicationInfoHash->Apply(ShutDownApplication, NULL);
 
-    // stop filewatcher monitoring thread
-    if (m_pFileWatcher != NULL)
-    {
-        delete  m_pFileWatcher;
-        m_pFileWatcher = NULL;
+            ReleaseSRWLockExclusive(&m_srwLock);
+        }
```
Comment (jkotalik) [inline src/AspNetCore/Src/applicationmanager.cxx]: clear the hashtable here?

File: src/AspNetCore/Src/applicationinfo.cpp [hunk 9]
```diff
@@ -146,10 +147,9 @@ APPLICATION_INFO::UpdateAppOfflineFileHandle()
                     strEventMsg.QueryStr());
             }
 
-            m_pApplication->ShutDown();
-            m_pApplication->DereferenceApplication();
+            // shut down application in another thread to not block FCN
+            APPLICATION_MANAGER::ShutDownApplication(this, NULL);
```
Comment (jkotalik) [inline src/AspNetCore/Src/applicationinfo.cpp]: We need to call recycle process for inprocess here.

File: src/AspNetCore/Src/applicationmanager.cxx (hunk 4 above)
Comment (pan-wang) [reply src/AspNetCore/Src/applicationmanager.cxx]: this method should be static. However, it depends on config and logically associated with a hashttable. Put is to into utility is not good for maintenance.

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 10]
```diff
@@ -189,80 +191,250 @@ APPLICATION_MANAGER::GetApplicationInfo(
...
     }
 }
+
+//
+// Function used by ShutDownApplication thread to do the real shutdown
```
Comment (pan-wang) [inline src/AspNetCore/Src/applicationmanager.cxx]: as the function got renamed, please update the comment

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 11]
```diff
@@ -189,80 +191,250 @@ APPLICATION_MANAGER::GetApplicationInfo(
...
-                ASPNETCORE_EVENT_RECYCLE_CONFIGURATION,
-                strEventMsg.QueryStr());
+            // When we are inprocess, we need to keep the application the
+            // application manager that is being deleted. This is because we will always need to recycle the worker
+            // process and any requests that hit this worker process must be rejected (while out of process can
+            // start a new dotnet process). We will immediately call Recycle after this call.
+
+            // We need to clear the hash table here and the ref copy the original table to avoid memory leaks
+            // and keep the application referenced.
+            m_pApplicationInfoHash->Clear();
```
Comment (pan-wang) [inline src/AspNetCore/Src/applicationmanager.cxx]: no need to call clear. For inprocess, m_pApplicationInfoHash should contain zero entry. You can do DBG_ASSERT(m_pApplicationInfoHash->Count() ==0)

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 12]
```diff
@@ -189,80 +191,250 @@ APPLICATION_MANAGER::GetApplicationInfo(
...
+
+            // shutdown will be done in another thread
+            RecycleApplication(pRecord, NULL);
+            pRecord->DereferenceApplicationInfo();
+            path = context.MultiSz.Next(path);
+        }
+    }
 
 Finished:
+    if (table != NULL && m_hostingModel != HOSTING_IN_PROCESS)
```
Comment (pan-wang) [inline src/AspNetCore/Src/applicationmanager.cxx]: only need to check table != NULL as already duplicated the entry back to m_pApplicationHaskTable. Otherwise, memory leak for the table

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 13]
```diff
@@ -189,80 +191,250 @@ APPLICATION_MANAGER::GetApplicationInfo(
...
+    // Null ref here.
+    pApplication->ReferenceApplication();
+
+    pEntry->ResetApplication();
+    pApplication->ShutDown();
+    pApplication->DereferenceApplication();
+}
+
+//
+// Function used to shutdown an application
```
Comment (pan-wang) [inline src/AspNetCore/Src/applicationmanager.cxx]: update comment

File: src/AspNetCore/Src/applicationinfo.cpp [hunk 14]
```diff
@@ -146,12 +155,22 @@ APPLICATION_INFO::UpdateAppOfflineFileHandle()
                     strEventMsg.QueryStr());
             }
 
-            m_pApplication->ShutDown();
-            m_pApplication->DereferenceApplication();
-            m_pApplication = NULL;
+            APPLICATION_MANAGER::GetInstance()->RecycleApplicationFromManager(
```
Comment (pan-wang) [inline src/AspNetCore/Src/applicationinfo.cpp]: you should call recycleapplication here which will just recycle specific application instance (as this is for app_offline) not for configuration change. Otherwise a subapplication may get recycled

File: src/AspNetCore/Src/applicationinfo.cpp [hunk 15]
```diff
@@ -146,12 +155,22 @@ APPLICATION_INFO::UpdateAppOfflineFileHandle()
...
+                m_pApplication->QueryConfig()->QueryConfigPath()->QueryStr()
+            );
 
+            if (m_pConfiguration->QueryHostingModel() == APP_HOSTING_MODEL::HOSTING_OUT_PROCESS)
```
Comment (pan-wang) [inline src/AspNetCore/Src/applicationinfo.cpp]: This can be put into RecycleApplication if you like

File: src/AspNetCore/Src/proxymodule.cxx [hunk 16]
```diff
@@ -128,10 +128,13 @@ ASPNET_CORE_PROXY_MODULE::OnExecuteRequestHandler(
             FALSE
         );
 
-        DataChunk.DataChunkType = HttpDataChunkFromMemory;
-        DataChunk.FromMemory.pBuffer = (PVOID)pAppOfflineHtm->m_Contents.QueryStr();
-        DataChunk.FromMemory.BufferLength = pAppOfflineHtm->m_Contents.QueryCB();
-        pResponse->WriteEntityChunkByReference(&DataChunk);
+        if (pAppOfflineHtm != NULL)
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/AspNetCore/Src/proxymodule.cxx]: There is already a debug assert now for AppOffline. I am just going to remove it.

File: src/CommonLib/aspnetcoreconfig.cxx [hunk 17]
```diff
@@ -46,12 +46,16 @@ ASPNETCORE_CONFIG::GetConfig(
     _In_  IHttpServer             *pHttpServer,
     _In_  HTTP_MODULE_ID           pModuleId,
     _In_  IHttpContext            *pHttpContext,
+    _In_  HANDLE                   hEventLog,
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/CommonLib/aspnetcoreconfig.cxx]: I don't believe that is available in the CommonLib library.

File: src/AspNetCore/Src/applicationinfo.cpp [hunk 18]
```diff
@@ -152,6 +160,12 @@ APPLICATION_INFO::UpdateAppOfflineFileHandle()
             m_pApplication = NULL;
         }
     }
+
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/AspNetCore/Src/applicationinfo.cpp]: I think this is fixed.

File: src/AspNetCore/Src/applicationinfo.cpp [hunk 19]
```diff
@@ -149,9 +158,14 @@ APPLICATION_INFO::UpdateAppOfflineFileHandle()
             m_pApplication->ShutDown();
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/AspNetCore/Src/applicationinfo.cpp]: Believe this is handled.

File: src/AspNetCore/Src/applicationmanager.cxx (hunk 12 above)
Comment (jkotalik) [reply src/AspNetCore/Src/applicationmanager.cxx]: Modifying to do a pointer copy instead of a table copy as I believe it will still work.

File: src/AspNetCore/Src/applicationinfo.cpp (hunk 15 above)
Comment (jkotalik) [reply src/AspNetCore/Src/applicationinfo.cpp]: Not moving this for now, but I think there a few potential issues with app_offline that I would like to address first.

File: src/CommonLib/resources.h [hunk 20]
```diff
@@ -28,8 +28,10 @@
 #define ASPNETCORE_EVENT_INPROCESS_THREAD_EXIT_STDERR_MSG    L"Application '%s' with physical root '%s' hit unexpected managed background thread exit, ErrorCode = '0x%x. First 4KB characters of captured stderr logs on startup:\r\n%s"
 #define ASPNETCORE_EVENT_INPROCESS_THREAD_EXIT_STDOUT_MSG    L"Application '%s' with physical root '%s' hit unexpected managed background thread exit, ErrorCode = '0x%x. Last 4KB characters of captured stdout and stderr logs:\r\n%s"
 #define ASPNETCORE_EVENT_INPROCESS_THREAD_EXIT_MSG           L"Application '%s' with physical root '%s' hit unexpected managed background thread exit, ErrorCode = '0x%x. Please check the stderr logs for more information."
-#define ASPNETCORE_EVENT_RECYCLE_APPOFFLINE_MSG              L"Application '%s' is recycled due to app_offline file was detected."
-#define ASPNETCORE_EVENT_RECYCLE_CONFIGURATION_MSG           L"Application '%s' is recycled due to configuration changed"
+#define ASPNETCORE_EVENT_APP_IN_SHUTDOWN_MSG                 L"Application shutting down."
+#define ASPNETCORE_EVENT_RECYCLE_APPOFFLINE_MSG              L"Application '%s' is recycled due to app_offline file."
```
Comment (Tratcher) [inline src/CommonLib/resources.h]: Are there separate messages for when the app_offline is added vs when it's removed? Is there a log for the application starting up in app_offline mode?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs [hunk 21]
```diff
@@ -199,6 +201,11 @@ public string ReasonPhrase
                 _reasonPhrase = value;
             }
         }
+        
+        public IISHttpServer Server
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs]: internal?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 22]
```diff
@@ -55,15 +60,50 @@ public Task StartAsync<TContext>(IHttpApplication<TContext> application, Cancell
 
         public Task StopAsync(CancellationToken cancellationToken)
         {
-            // TODO: Drain pending requests
+            // First call back into native saying "DON'T SEND ME ANY MORE REQUESTS"
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: This comment is in the wrong place, move it down

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 23]
```diff
@@ -55,15 +60,50 @@ public Task StartAsync<TContext>(IHttpApplication<TContext> application, Cancell
...
+            void RegisterCancelation()
+            {
+                cancellationToken.Register(() =>
+                {
+                    _shutdownSignal.TrySetResult(null);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: Does this need to call `lNativeMethods.http_stop_calls_into_mananged()`? Or will that be handled somewhere else in the timeout scenario?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 24]
```diff
@@ -55,15 +60,50 @@ public Task StartAsync<TContext>(IHttpApplication<TContext> application, Cancell
...
+            {
+                _shutdownSignal.TrySetException(ex);
+            }
+
+            return _shutdownSignal.Task;
         }
 
         public void Dispose()
         {
+            // This can get triggered too early...
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: too early?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 25]
```diff
@@ -55,15 +60,50 @@ public Task StartAsync<TContext>(IHttpApplication<TContext> application, Cancell
...
+            _stopping = 1;
+            _shutdownSignal.TrySetResult(null);
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: What about NativeMethods.http_stop_calls_into_mananged(); and http_stop_incoming_requests()?

File: src/CommonLib/resources.h (hunk 20 above)
Comment (jkotalik) [reply src/CommonLib/resources.h]: There is currently no log for when app_offline is removed. I will add one now.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs (hunk 24 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: sorry, obsolete comment.

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs (hunk 25 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: I can just set http_stop_incoming_requests().

Comment (Tratcher) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: Why? Dispose may be called without StopAsync for error cases. What's to prevent requests in progress from AV'ing without stop_calls?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 26]
```diff
@@ -183,6 +187,12 @@ public IISHttpContext CreateHttpContext(IntPtr pInProcessHandler)
                 return new IISHttpContextOfT<T>(_memoryPool, _application, pInProcessHandler, _options, _server);
             }
         }
+
+        ~IISHttpServer()
+        {
+            // If this finalize is invoked, try our best to block all requests into managed.
+            NativeMethods.http_stop_incoming_requests();
```
Comment (Tratcher) [inline src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: What about stop_calls?

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs (hunk 25 above)
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: Sorry I meant NativeMethods.http_stop_calls_into_managed(). 😨

File: src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs [hunk 27]
```diff
@@ -105,6 +150,13 @@ private static void CompleteRequest(IISHttpContext context, Task<bool> completed
             // Post completion after completing the request to resume the state machine
             context.PostCompletion(ConvertRequestCompletionResults(completedTask.Result));
 
+            if (Interlocked.Decrement(ref context.Server._outstandingRequests) == 0 && context.Server.Stopping)
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpServer.cs]: CompleteRequest will not be skipped and PostCompletion will never fail AFAIK. @pan-wang PostCompletion can never fail, based on this: 
https://msdn.microsoft.com/en-us/library/ms692997(v=vs.90).aspx

File: src/RequestHandler/inprocess/inprocessapplication.cpp [hunk 28]
```diff
@@ -160,27 +221,47 @@ IN_PROCESS_APPLICATION::OnExecuteRequest(
...
+        }
+
+        pHttpContext->GetResponse()->SetStatus(500,
+            "Internal Server Error",
+            0,
             (ULONG)E_APPLICATION_ACTIVATION_EXEC_FAILURE);
+
+        dwRequestNotificationStatus = RQ_NOTIFICATION_FINISH_REQUEST;
+    }
+    else if (m_status != APPLICATION_STATUS::RUNNING || m_fBlockCallbacksIntoManaged)
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/RequestHandler/inprocess/inprocessapplication.cpp]: Would it? m_fBlockCallbacksIntoManaged is volatile so it would be atomic.

[the remark this answers is not in the document]
Comment (Tratcher) [reply src/RequestHandler/inprocess/inprocessapplication.cpp]: You have a (small) gap between checking the bool and calling the delegate.

File: src/AspNetCore/Src/applicationmanager.cxx [hunk 29]
```diff
@@ -189,80 +201,306 @@ APPLICATION_MANAGER::GetApplicationInfo(
...
+            // When we are inprocess, we need to keep the application the
+            // application manager that is being deleted. This is because we will always need to recycle the worker
+            // process and any requests that hit this worker process must be rejected (while out of process can
+            // start a new dotnet process). We will immediately call Recycle after this call.
+            DBG_ASSERT(m_pApplicationInfoHash->Count() == 0);
+            delete m_pApplicationInfoHash;
+
+            // We don't want to delete the table as m_pApplicationInfoHash = table
+            fKeepTable = TRUE;
+            m_pApplicationInfoHash = table;
```
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/AspNetCore/Src/applicationmanager.cxx]: I deleted m_pApplicationInfoHash right before this...

**Comment by pan-wang** [comment]
Let's commit this one first and fix the issues parallel instead of holding it for long time

File: src/RequestHandler/inprocess/inprocessapplication.cpp (hunk 28 above)
[the remark this answers is not in the document]
Comment (jkotalik) [reply src/RequestHandler/inprocess/inprocessapplication.cpp]: It's still the same race, except worse because we are guaranteed to AV. We still need to check if the handler is null or not.
